main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1715 @@
+    def apply(l: List[Equation]): List[Equation] = l.map(apply)
+
+    /**
+      * Returns the number of bindings in `this` substitution.
+      */
+    def numberOfBindings: Int = m.size
+
+    /**
+      * Returns the size of `this` substitution.
+      *
+      * The size of a substitution is the sum of the sizes of all terms in its co-domain.
+      */
+    def size: Int = m.values.map(_.size).sum
+
+    /**
+      * Extends `this` substitution with a new binding from the variable `x` to the term `t`.
+      *
+      * If the variable `x` already occurs in `this` substitution then it must be bound to the same term as `t`.
+      */
+    def extended(x: Int, t: Term, loc: SourceLocation): BoolSubstitution = m.get(x) match {
+      case None => BoolSubstitution(m + (x -> t))
+      case Some(t1) =>
+        // Note: If t == t1 we can just return the same substitution.
+        if (t == t1)
+          this
+        else
+          throw ConflictException(t, t1, loc)
+    }
+
+    /**
+      * Merges `this` substitution with `that` substitution.
+      *
+      * The returned substitution has all bindings from `this` and `that` substitution.
+      *
+      * The domains of the two substitutions must not overlap.
+      */
+    def ++(that: BoolSubstitution): BoolSubstitution = {
+      val intersection = this.m.keySet.intersect(that.m.keySet)
+      if (intersection.nonEmpty) {
+        throw InternalCompilerException(s"Substitutions are not disjoint on: '${intersection.mkString(",")}'.", SourceLocation.Unknown)
+      }
+
+      BoolSubstitution(this.m ++ that.m)
+    }
+
+    /**
+      * Composes `this` substitution with `that` substitution.
+      *
+      * Conceptually `this` is a function: `b -> c` and `that` is a function `a -> b`.
+      *
+      * We want to compute `a -> c` which we get by computing `x -> this(that(x))`.
+      */
+    def @@(that: BoolSubstitution): BoolSubstitution = {
+      // Case 1: Return `that` if `this` is empty.
+      if (this.m.isEmpty) {
+        return that
+      }
+
+      // Case 2: Return `this` if `that` is empty.
+      if (that.m.isEmpty) {
+        return this
+      }
+
+      // Case 3: Merge the two substitutions.
+      val result = mutable.Map.empty[Int, Term]
+
+      // Add all bindings in `that`. (Applying the current substitution).
+      for ((x, t) <- that.m) {
+        result.update(x, this.apply(t))
+      }
+
+      // Add all bindings in `this` that are not in `that`.
+      for ((x, t) <- this.m) {
+        if (!that.m.contains(x)) {
+          result.update(x, t)
+        }
+      }
+
+      BoolSubstitution(result.toMap)
+    }
+
+    /**
+      * Returns a human-readable representation of `this` substitution.
+      */
+    override def toString: String = {
+      val indent = 4
+
+      val sb = new StringBuilder()
+      // We sort the bindings by (size, name).
+      for ((x, t) <- m.toList.sortBy(kv => (kv._2.size, kv._1))) {
+        sb.append(" ".repeat(indent))
```

Comment:
Is it so costly to map and mkString?

Reply:
If we are using toString its probably because something has gone wrong and the result is big. Thats why I use a string builder etc.